### DISCUSSION TITLE
Speed up response time

### DISCUSSION THREAD

**Discussion by IRediTOTO** [wp-graphql/wp-graphql#1813] [General] [answered]
Hello, can you please make response time faster? I tried in a blank WP with 10 products but time response up to 500ms :( 
This too longgggg, I think. 
I tried plugin API cache WP REST Cache but look like it not work with wp graphql. 
Thank you

**Comment by izzygld** [comment]
@warcraft14115 not sure how you are implementing your data,

have you tried apollo cache?

take a look at [https://www.apollographql.com/blog/graphql-caching-the-elephant-in-the-room-11a3df0c23ad](https://www.apollographql.com/blog/graphql-caching-the-elephant-in-the-room-11a3df0c23ad)

**Comment by CodeProKid** [comment]
@warcraft14115 can you give some more info on the query you are making that is taking this long? Could you also let us know what version of the plugin you are on, and how you are making this query (ex. from the client, through WPGraphiQL, From another server, etc...).

**Comment by dpacmittal** [comment]
@CodeProKid Is there a hook/filter which runs right before returning the response? I'd like to cache queries using transients api. I've looked all over docs and github but couldn't find anything.

**Comment by dpacmittal** [comment]
Nvm, I just found out about `graphql_execute` and `graphql_request_results`. Now just need to find a way to short circuit the request.

**Comment by izzygld** [comment]
@dpacmittal did you ever try working on using `graphql_execute` and `graphql_request_results` to cache queries using transients api?

**Comment by dpacmittal** [comment]
@izzygld I am using `do_graphql_request` and `graphql_return_response` hooks and it works great. I'm not using transients api though, just storing in redis.

**Comment by izzygld** [comment]
can you share an example of how you are doing it?

Are you wrapping the whole wpgraghql into a `do_graphql_request` or just specific parts?

**Comment by dpacmittal** [comment, marked as answer]
@izzygld You should use https://github.com/valu-digital/wp-graphql-cache. If you still want to use a custom solution (like I did because I wanted to do stale-while-revalidate), here it is - https://gist.github.com/dpacmittal/e837e2a36714c6d1a279fbd5999debd0

**Reply by metawise** [reply]
I tried this and it worked. Went down from ~750ms to 400ms. But still, meanwhile REST API return the same results in 50ms.

**Comment by jasonbahl** [comment]
I'm going to transfer this to a Discussion. I think it's a good discussion and a lot of ways we can go about caching requests, but not an actionable issue on its own.

**Comment by metawise** [comment]
I wonder if there is any improvements on this as far as speed goes?

**Comment by jasonbahl** [comment]
I wanted to follow up here as WPGraphQL v1.6.0 made a lot of performance improvements, that should help with most requests: 

See: https://github.com/wp-graphql/wp-graphql/releases/tag/v1.6.0

The best way to speed things up, however, is to reduce how much needs to execute in the first place. 

Making GET requests can help here, as it can take advantage of HTTP caching and never even hit the WordPress application layer. 

For example: 

![Screen Shot 2021-12-10 at 9 05 21 AM](https://user-images.githubusercontent.com/1260765/145604883-106d49c1-87b9-4646-ad3b-bc60ecdfceef.png)

In the screenshot we can see a GET request for a GraphQL query returning in `56ms`. 

This is faster than a REST request also for 10 posts on the same server, also using GET, also with Cache Control set. 

![Screen Shot 2021-12-10 at 9 11 05 AM](https://user-images.githubusercontent.com/1260765/145605376-77ab1052-3a4c-419c-be61-0757a602780f.png)

Often, WPGraphQL will be faster than REST because there's less data to download. 

But if you're comparing GET requests to a REST endpoint with POST requests to a GraphQL endpoint, then yes, GraphQL will be slower because you're always making the full trip to the server, and not able to take advantage of the HTTP caching. 

Solutions like WP GraphQL Cache from Valu Digital are a good step toward making POST requests respond faster as it eliminates a lot of underlying execution. 

We're working on some things in this area that should benefit all WPGraphQL users in the near future, so stay tuned for updates!

**Comment by wildberry-source** [comment]
will redis caching work?

**Comment by jasonbahl** [comment]
I wanted to follow-up again on this discussion. We have https://github.com/wp-graphql/wp-graphql/pull/3356 in progress which is showing some good improvements, specifically around reducing the number of strings that are translated during WPGraphQL request execution (i.e. descriptions and deprecation reasons that are only needed for introspection). 

Would love to see if anyone that's had performance issues would be willing to test that branch and provide feedback: https://github.com/wp-graphql/wp-graphql/pull/3356
